components/cloud/src/blob.rs

```diff
@@ -0,0 +1,191 @@
+use futures_io::AsyncRead;
+pub use kvproto::backup::CloudDynamic;
+use std::io;
+use std::marker::Unpin;
+
+/// An abstraction for blob storage.
+/// Currently the same as ExternalStorage
+pub trait BlobStorage: 'static + Send + Sync {
+    fn name(&self) -> &'static str;
+
+    fn url(&self) -> io::Result<url::Url>;
+
+    /// Write all contents of the read to the given path.
+    fn write(
+        &self,
+        name: &str,
+        reader: Box<dyn AsyncRead + Send + Unpin>,
+        content_length: u64,
+    ) -> io::Result<()>;
+
+    /// Read all contents of the given path.
+    fn read(&self, name: &str) -> Box<dyn AsyncRead + Unpin + '_>;
+}
+
+impl BlobStorage for Box<dyn BlobStorage> {
+    fn name(&self) -> &'static str {
+        (**self).name()
+    }
+
+    fn url(&self) -> io::Result<url::Url> {
+        (**self).url()
+    }
+
+    fn write(
```

Comment:
put_object/get_object is more of "blob storage" than read/write?

Reply:
Yes. Currently there is no difference with ExternalStorage so it made sense to keep the name the same until the API actually changes to better match blob storage.

Reply:
but "read/write" is confusing in the context of blob storage, if someday someone decide to use the blob storage API somewhere else.

Reply:
Okay, I renamed to get/put